File: db4oj/core/src/com/db4o/inside/btree/BTreePointer.java
/* Copyright (C) 2006  db4objects Inc.  http://www.db4o.com */

package com.db4o.inside.btree;

import com.db4o.*;
import com.db4o.foundation.*;
import com.db4o.inside.*;

/**
 * @exclude
 */
public class BTreePointer{
	
	public static BTreePointer max(BTreePointer x, BTreePointer y) {
		if (x == null) {
			return x;
		}
		if (y == null) {
			return y;
		}
		if (x.compareTo(y) > 0) {
			return x;
		}
		return y;
	}

	public static BTreePointer min(BTreePointer x, BTreePointer y) {
		if (x == null) {
			return y;
		}
		if (y == null) {
			return x;
		}
		if (x.compareTo(y) < 0) {
			return x;
		}
		return y;
	}
    
    private final BTreeNode _node;
    
    private final int _index;

	private final Transaction _transaction;

	private final Buffer _nodeReader;
   
    public BTreePointer(Transaction transaction, Buffer nodeReader, BTreeNode node, int index) {
    	if(transaction == null || node == null){
            throw new ArgumentNullException();
        }
        _transaction = transaction;
        _nodeReader = nodeReader;
        _node = node;
        _index = index;
	}

	public final Transaction transaction() {
    	return _transaction;
    }
    
    public final int index(){
        return _index;
    }
    
    public final BTreeNode node() {
        return _node;
    }
    
    public final Object key() {
		return node().key(transaction(), nodeReader(), index());
	}

	public final Object value() {
		return node().value(nodeReader(), index());
	}
	
	private Buffer nodeReader() {
		return _nodeReader;
	}
    
    public BTreePointer next(){
        int indexInMyNode = index() + 1;
        while(indexInMyNode < node().count()){
            if(node().indexIsValid(transaction(), indexInMyNode)){
                return new BTreePointer(transaction(), nodeReader(), node(), indexInMyNode);
            }
            indexInMyNode ++;
        }
        int newIndex = -1;
        BTreeNode nextNode = node();
        Buffer nextReader = null;
        while(newIndex == -1){
            nextNode = nextNode.nextNode();
            if(nextNode == null){
                return null;
            }
            nextReader = nextNode.prepareRead(transaction());
            newIndex = nextNode.firstKeyIndex(transaction());
        }
        return new BTreePointer(transaction(), nextReader, nextNode, newIndex);
    }
    
	public BTreePointer previous() {
		int indexInMyNode = index() - 1;
		while(indexInMyNode >= 0){
			if(node().indexIsValid(transaction(), indexInMyNode)){
				return new BTreePointer(transaction(), nodeReader(), node(), indexInMyNode);
			}
			indexInMyNode --;
		}
		int newIndex = -1;
		BTreeNode previousNode = node();
		Buffer previousReader = null;
		while(newIndex == -1){
			previousNode = previousNode.previousNode();
			if(previousNode == null){
				return null;
			}
			previousReader = previousNode.prepareRead(transaction());
			newIndex = previousNode.lastKeyIndex(transaction());
		}
		return new BTreePointer(transaction(), previousReader, previousNode, newIndex);
	}    

    
    public boolean equals(Object obj) {
        if(this == obj){
            return true;
        }
        if(! (obj instanceof BTreePointer)){
            return false;
        }
        BTreePointer other = (BTreePointer) obj;
        
        if(index() != other.index()){
            return false;
        }
        
        return node().equals(other.node());
    }	
    
    public int hashCode() {
    	return node().hashCode();
    }
    
    public String toString() {
        return "BTreePointer(index=" + index() + ", node=" + node() + ")";      
    }

	public int compareTo(BTreePointer y) {
		if (null == y) {
			throw new ArgumentNullException();
		}
		if (btree() != y.btree()) {
			throw new IllegalArgumentException();
		}		
		return btree().compareKeys(key(), y.key());
	}

	private BTree btree() {
		return node().btree();
	}

	public static boolean lessThan(BTreePointer x, BTreePointer y) {
		return BTreePointer.min(x, y) == x
			&& !equals(x, y);
	}

	public static boolean equals(BTreePointer x, BTreePointer y) {
		if (x == null) {
			return y == null;
		}
		return x.equals(y);
	}

	public boolean isValid() {
		return node().indexIsValid(transaction(), index());
	}    
}

File: db4on/core/src/core/com/db4o/inside/btree/BTreePointer.cs
namespace com.db4o.inside.btree
{
	/// <exclude></exclude>
	public class BTreePointer
	{
		public static com.db4o.inside.btree.BTreePointer Max(com.db4o.inside.btree.BTreePointer
			 x, com.db4o.inside.btree.BTreePointer y)
		{
			if (x == null)
			{
				return x;
			}
			if (y == null)
			{
				return y;
			}
			if (x.CompareTo(y) > 0)
			{
				return x;
			}
			return y;
		}

		public static com.db4o.inside.btree.BTreePointer Min(com.db4o.inside.btree.BTreePointer
			 x, com.db4o.inside.btree.BTreePointer y)
		{
			if (x == null)
			{
				return y;
			}
			if (y == null)
			{
				return x;
			}
			if (x.CompareTo(y) < 0)
			{
				return x;
			}
			return y;
		}

		private readonly com.db4o.inside.btree.BTreeNode _node;

		private readonly int _index;

		private readonly com.db4o.Transaction _transaction;

		private readonly com.db4o.YapReader _nodeReader;

		public BTreePointer(com.db4o.Transaction transaction, com.db4o.YapReader nodeReader
			, com.db4o.inside.btree.BTreeNode node, int index)
		{
			if (transaction == null || node == null)
			{
				throw new System.ArgumentNullException();
			}
			_transaction = transaction;
			_nodeReader = nodeReader;
			_node = node;
			_index = index;
		}

		public com.db4o.Transaction Transaction()
		{
			return _transaction;
		}

		public int Index()
		{
			return _index;
		}

		public com.db4o.inside.btree.BTreeNode Node()
		{
			return _node;
		}

		public object Key()
		{
			return Node().Key(Transaction(), NodeReader(), Index());
		}

		public object Value()
		{
			return Node().Value(NodeReader(), Index());
		}

		private com.db4o.YapReader NodeReader()
		{
			return _nodeReader;
		}

		public virtual com.db4o.inside.btree.BTreePointer Next()
		{
			int indexInMyNode = Index() + 1;
			while (indexInMyNode < Node().Count())
			{
				if (Node().IndexIsValid(Transaction(), indexInMyNode))
				{
					return new com.db4o.inside.btree.BTreePointer(Transaction(), NodeReader(), Node()
						, indexInMyNode);
				}
				indexInMyNode++;
			}
			int newIndex = -1;
			com.db4o.inside.btree.BTreeNode nextNode = Node();
			com.db4o.YapReader nextReader = null;
			while (newIndex == -1)
			{
				nextNode = nextNode.NextNode();
				if (nextNode == null)
				{
					return null;
				}
				nextReader = nextNode.PrepareRead(Transaction());
				newIndex = nextNode.FirstKeyIndex(Transaction());
			}
			return new com.db4o.inside.btree.BTreePointer(Transaction(), nextReader, nextNode
				, newIndex);
		}

		public override bool Equals(object obj)
		{
			if (this == obj)
			{
				return true;
			}
			if (!(obj is com.db4o.inside.btree.BTreePointer))
			{
				return false;
			}
			com.db4o.inside.btree.BTreePointer other = (com.db4o.inside.btree.BTreePointer)obj;
			if (Index() != other.Index())
			{
				return false;
			}
			return Node().Equals(other.Node());
		}

		public override string ToString()
		{
			return "BTreePointer(index=" + Index() + ", node=" + Node() + ")";
		}

		public virtual int CompareTo(com.db4o.inside.btree.BTreePointer y)
		{
			if (null == y)
			{
				throw new System.ArgumentNullException();
			}
			if (Btree() != y.Btree())
			{
				throw new System.ArgumentException();
			}
			return Btree().CompareKeys(Key(), y.Key());
		}

		private com.db4o.inside.btree.BTree Btree()
		{
			return Node().Btree();
		}

		public static bool LessThan(com.db4o.inside.btree.BTreePointer x, com.db4o.inside.btree.BTreePointer
			 y)
		{
			return com.db4o.inside.btree.BTreePointer.Min(x, y) == x && !Equals(x, y);
		}

		public static bool Equals(com.db4o.inside.btree.BTreePointer x, com.db4o.inside.btree.BTreePointer
			 y)
		{
			if (x == null)
			{
				return y == null;
			}
			return x.Equals(y);
		}

		public virtual bool IsValid()
		{
			return Node().IndexIsValid(Transaction(), Index());
		}
	}
}
